core/Views/ViewManager.vala

```diff
@@ -0,0 +1,213 @@
+
+    /**
+    * The visible view
+    */
+    public View selected_view { get; private construct set; }
+
+    /**
+    * A new view have been registered
+    */
+    public signal void view_added (View new_view);
+
+    /**
+    * A view have been unregistered
+    */
+    public signal void view_removed (View view);
+
+    /**
+    * A category have been registered
+    */
+    public signal void category_added (Category cat);
+
+    /**
+    * A category have been unregistered
+    */
+    public signal void category_removed (Category cat);
+
+    /**
+    * A view wants to be refiltered
+    */
+    public signal void request_filtering (View view);
+
+    construct {
+        views = new Gee.ArrayList<View> ();
+        categories = new Gee.ArrayList<Category> ();
+    }
+
+    /**
+    * Register a new category
+    */
+    public void add_category (Category cat) {
+        if (!(cat in categories)) {
+            categories.add (cat);
+            category_added (cat);
+            cat.remove.connect (() => {
+                category_removed (cat);
+            });
+        }
+    }
+
+    /**
+    * Try to retrieve a registered category.
+    *
+    * @return The requested category if found
+    * @throw Throws an error if the category couldn't be found
+    */
+    public Category get_category (string id) throws ViewError {
+        foreach (var cat in categories) {
+            if (cat.id == id) {
+                return cat;
+            }
+        }
+
+        throw new ViewError.NOT_FOUND (@"Cant't find any category with the following ID: $id");
+    }
+
+    /**
+    * Register a new view.
+    *
+    * Make sure you have registered its category before.
+    */
+    public void add (View view, bool select = false) {
+        try {
+            get_category (view.category);
+            if (!(view in views)) {
+                debug ("Adding view: %s", view.id);
+                views.add (view);
+
+                if (selected_view == null || select) {
+                    selected_view = view;
+                }
+
+                view_added (view);
+
+                view.request_filtering.connect (() => {
+                    request_filtering (view);
+                });
+
+                view.remove.connect(() => {
+                    view_removed (view);
+                });
+            }
+        } catch (Error err) {
+            warning (@"Tried to register a new view ($(view.id)) but its category ($(view.category)) wasn't registered yet.");
+        }
+    }
+
+    /**
+    * Remove a view.
+    */
+    public void remove_view (View view) {
+        views.remove (view);
+        view.destroy ();
+        view_removed (view);
+    }
+
+    /**
+    * Try to retrieve a registered view.
+    *
+    * @return The requested view if found
+    * @throw Throws an error if the view couldn't be found
+    */
+    public new View get (string id) throws ViewError {
```

Comment:
This is not good, don't use the `new` attribute, it's hiding things and leads to mistakes, better rename that function